File: android/src/main/kotlin/com/gtbluesky/fusion/notification/FusionNotification.kt
package com.gtbluesky.fusion.notification

import java.lang.ref.WeakReference

interface FusionNotificationListener {
    fun onReceive(name: String, body: Map<String, Any>?)
}

object FusionNotificationBinding {
    private val listeners = mutableListOf<WeakReference<FusionNotificationListener>>()

    fun register(listener: FusionNotificationListener) {
        unregister(listener)
        listeners.add(WeakReference(listener))
    }

    fun unregister(listener: FusionNotificationListener) {
        listeners.removeAll {
            it.get() == listener || it.get() == null
        }
    }

    internal fun dispatchMessage(name: String, body: Map<String, Any>?) {
        listeners.forEach {
            it.get()?.onReceive(name, body)
        }
    }
}

enum class FusionNotificationType {
    FLUTTER,
    NATIVE,
    GLOBAL
}

File: ios/Classes/Notification/FusionNotification.swift
//
//  FusionNotification.swift
//  fusion
//
// Created by gtbluesky on 2022/4/28.
//

import Foundation

@objc public protocol FusionNotificationListener {
    func onReceive(name: String, body: Dictionary<String, Any>?)
}

@objc public class FusionNotificationBinding: NSObject {
    private var listeners = [WeakReference<FusionNotificationListener>]()
    public static let instance = FusionNotificationBinding()

    private override init() {
        super.init()
    }

    public func register(_ listener: FusionNotificationListener) {
        unregister(listener)
        listeners.append(WeakReference(listener))
    }

    public func unregister(_ listener: FusionNotificationListener) {
        listeners.removeAll {
            ($0.value as AnyObject) === (listener as AnyObject) || $0.value == nil
        }
    }

    internal func dispatchMessage(_ name: String, _ body: Dictionary<String, Any>?) {
        listeners.forEach {
            $0.value?.onReceive(name: name, body: body)
        }
    }
}

@objc public enum FusionNotificationType: Int {
    case flutter
    case native
    case global
  }
